localstack/services/events/provider_v2.py

```diff
@@ -1,30 +1,95 @@
+import base64
 import logging
+from typing import Optional
 
 from localstack.aws.api import RequestContext, handler
 from localstack.aws.api.events import (
+    Arn,
+    Boolean,
     CreateEventBusResponse,
+    DescribeEventBusResponse,
+    DescribeRuleResponse,
+    EndpointId,
+    EventBusList,
     EventBusName,
     EventBusNameOrArn,
     EventPattern,
     EventsApi,
     EventSourceName,
+    LimitMax100,
+    ListEventBusesResponse,
+    ListRuleNamesByTargetResponse,
+    ListRulesResponse,
+    ListTargetsByRuleResponse,
+    NextToken,
+    PutEventsRequestEntryList,
+    PutEventsResponse,
+    PutPartnerEventsRequestEntryList,
+    PutPartnerEventsResponse,
     PutRuleResponse,
+    PutTargetsResponse,
+    RemoveTargetsResponse,
+    ResourceAlreadyExistsException,
+    ResourceNotFoundException,
     RoleArn,
     RuleDescription,
     RuleName,
+    RuleResponseList,
     RuleState,
     ScheduleExpression,
     TagList,
+    Target,
+    TargetArn,
+    TargetId,
+    TargetIdList,
+    TargetList,
 )
+from localstack.aws.api.events import EventBus as ApiTypeEventBus
+from localstack.aws.api.events import Rule as ApiTypeRule
+from localstack.services.events.event_bus import EventBusService, EventBusServiceDict
+from localstack.services.events.models_v2 import (
+    EventBus,
+    EventBusDict,
+    EventsStore,
+    Rule,
+    RuleDict,
+    TargetDict,
+    ValidationException,
+    events_store,
+)
+from localstack.services.events.rule import RuleService, RuleServiceDict
+from localstack.services.events.target import TargetService, TargetServiceDict, TargetServiceFactory
 from localstack.services.plugins import ServiceLifecycleHook
 
 LOG = logging.getLogger(__name__)
 
 
+def decode_next_token(token: NextToken) -> int:
+    """Decode a pagination token from base64 to integer."""
+    return int.from_bytes(base64.b64decode(token), "big")
+
+
+def encode_next_token(token: int) -> NextToken:
+    """Encode a pagination token to base64 from integer."""
+    return base64.b64encode(token.to_bytes(128, "big")).decode("utf-8")
+
+
+def get_filtered_dict(name_prefix: str, input_dict: dict) -> dict:
+    """Filter dictionary by prefix."""
+    return {name: value for name, value in input_dict.items() if name.startswith(name_prefix)}
+
+
 class EventsProvider(EventsApi, ServiceLifecycleHook):
+    # api methods are grouped by resource type and sorted in hierarchical order
+    # each group is sorted alphabetically
     def __init__(self):
-        self._rules = {}
-        self._event_buses = {}
+        self._event_bus_services_store: EventBusServiceDict = {}
+        self._rule_services_store: RuleServiceDict = {}
```

Comment:
Yes, this is mostly preparation for the future implementation dealing with persistence. My goal is to add an additional abstraction on top of the AccountRegionBundle, that combines state that is persistent and logic into an object, so that the provider only sees this object.

Reply:
As discussed, the question was less related to the necessity of the `...Service` objects themselves, but whether we need to store them if there is no state in them. We can keep it for now, and see in a future iteration how they will be handled.